extensions/controllers/sandboxclaim_controller.go

```diff
@@ -76,26 +81,64 @@ func (r *SandboxClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, nil
 	}
 
-	// cache the original status from sandboxclaim
+	// Cache original status
 	originalClaimStatus := claim.Status.DeepCopy()
+
 	var err error
 	var sandbox *v1alpha1.Sandbox
 	var template *extensionsv1alpha1.SandboxTemplate
 
-	// Try getting template
-	if template, err = r.getTemplate(ctx, claim); err == nil || k8errors.IsNotFound(err) {
-		// This ensures the firewall is up before the pod starts.
-		if npErr := r.reconcileNetworkPolicy(ctx, claim, template); npErr != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to reconcile network policy: %w", npErr)
+	// Check Expiration
+	claimExpired := false
+	if claim.Spec.Lifecycle != nil && claim.Spec.Lifecycle.ShutdownTime != nil {
+		claimExpired = time.Now().After(claim.Spec.Lifecycle.ShutdownTime.Time)
+	}
+
+	// 1. Manage Sandbox Resources
+	if !claimExpired {
+		// Logic for active Claims
+		if template, err = r.getTemplate(ctx, claim); err == nil || k8errors.IsNotFound(err) {
+			// Ensure NetworkPolicy
+			if npErr := r.reconcileNetworkPolicy(ctx, claim, template); npErr != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to reconcile network policy: %w", npErr)
+			}
+			sandbox, err = r.getOrCreateSandbox(ctx, claim, template)
+		}
+	} else {
+		// Logic for expired Claims (Only check existence, do not create)
+		existingSandbox := &v1alpha1.Sandbox{}
+		if getErr := r.Get(ctx, req.NamespacedName, existingSandbox); getErr == nil {
+			sandbox = existingSandbox
+		}
+	}
+
+	// 2. Logic for Auto-Deletion (Policy=Delete)
+	// We delete if: No error occurred AND Policy is Delete AND (Sandbox is missing/expired)
+	if err == nil && claim.Spec.Lifecycle != nil && claim.Spec.Lifecycle.ShutdownPolicy == extensionsv1alpha1.ShutdownPolicyDelete {
+		resourceExpired := (sandbox == nil && claimExpired) || (sandbox != nil && isSandboxExpired(sandbox))
```

Comment:
This implies that claim deletion post expiration is blocked by sandbox expiration.

Reply:
You are right coupling the Claim deletion to the Sandbox status was unnecessary and in fact introduced a potential race condition (recreation loop).

I have refactored this to treat the Claim as the source of truth. If the Claim is expired and the policy is Delete, the controller now deletes the Claim immediately without waiting for the Sandbox to reflect that state.